### PR TITLE
KAFKA-17233: MirrorCheckpointConnector should use batched listConsumerGroupOffsets

### PR DESCRIPTION
**Pull request by fonsdant**
Of the Admin interface, the listConsumerGroupOffsets by groupSpecs was chosen
because it is the most similar method to the one by groupId. Both methods do
not need to specify options, but create them with the
ListConsumerGroupOffsetsOptions empty constructor and both end up calling
listConsumerGroupOffsets with a created groupSpecs map and default options.

For this reason, the method listConsumerGroupOffsets of class
MirrorCheckpointConnector needed to be changed to adjust itself to the new
method used.

Since all group offsets are now obtained asynchronously, it is necessary to 
modify the for loop in findConsumerGroups: the ListConsumerGroupOffsetsResult
must be obtained outside of the for loop, and for each group it is needed to get
its topic partition and metadata.

To adjust the tests and perform the mock for listConsumerGroupOffsets, it has
needed to create futures to create a ListConsumerGroupOffsetsResult as the
return of listConsumerGroupOffsets. For this reason, the empty constructor has
become public to be accessed inside the test class.

The assertEquals argument order has been fixed as respectively expected and
actual.

The offsetsForGroup4 instance has been removed as listConsumerGroupOffsets could
not return it as part of the result. This is because group "g4" would be
filtered and not be in the filteredGroup that is passed to
listConsumerGroupOffsets.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by fonsdant** [comment]
@gharris1727, could you check it?

File: clients/src/main/java/org/apache/kafka/clients/admin/ListConsumerGroupOffsetsResult.java [hunk 1]
```diff
@@ -40,7 +40,7 @@ public class ListConsumerGroupOffsetsResult {
 
     final Map<String, KafkaFuture<Map<TopicPartition, OffsetAndMetadata>>> futures;
 
-    ListConsumerGroupOffsetsResult(final Map<CoordinatorKey, KafkaFuture<Map<TopicPartition, OffsetAndMetadata>>> futures) {
+    public ListConsumerGroupOffsetsResult(final Map<CoordinatorKey, KafkaFuture<Map<TopicPartition, OffsetAndMetadata>>> futures) {
```
Comment (gharris1727) [inline clients/src/main/java/org/apache/kafka/clients/admin/ListConsumerGroupOffsetsResult.java]: This is a change to the public interface, which would require a KIP. Please revert this change and find an alternative. Thanks!

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnector.java [hunk 2]
```diff
@@ -262,12 +264,13 @@ private void createInternalTopics() {
         );
     }
 
-    Map<TopicPartition, OffsetAndMetadata> listConsumerGroupOffsets(String group)
+    ListConsumerGroupOffsetsResult listConsumerGroupOffsets(List<String> groups)
             throws InterruptedException, ExecutionException {
+        Map<String, ListConsumerGroupOffsetsSpec> groupSpecs = groups.stream()
+                .collect(Collectors.toMap(group -> group, group -> new ListConsumerGroupOffsetsSpec()));
```
Comment (gharris1727) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnector.java]: nit: create one of these Spec objects and reuse it for every entry in the map.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnectorTest.java [hunk 3]
```diff
@@ -151,7 +155,14 @@ public void testFindConsumerGroups() throws Exception {
         doReturn(groups).when(connector).listConsumerGroups();
         doReturn(true).when(connector).shouldReplicateByTopicFilter(anyString());
         doReturn(true).when(connector).shouldReplicateByGroupFilter(anyString());
-        doReturn(offsets).when(connector).listConsumerGroupOffsets(anyString());
+
+        Map<CoordinatorKey, KafkaFuture<Map<TopicPartition, OffsetAndMetadata>>> futures = new HashMap<>();
+        futures.put(CoordinatorKey.byGroupId("g1"), KafkaFuture.completedFuture(offsets));
+        futures.put(CoordinatorKey.byGroupId("g2"), KafkaFuture.completedFuture(offsets));
+        ListConsumerGroupOffsetsResult offsetsResult = new ListConsumerGroupOffsetsResult(futures);
+        offsetsResult = spy(offsetsResult);
```
Comment (gharris1727) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnectorTest.java]: I think you can avoid needing to change the constructor visibility if you use mock(ListConsumerGroupOffsetsResult.class) instead.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnectorTest.java [hunk 4]
```diff
@@ -151,7 +155,14 @@ public void testFindConsumerGroups() throws Exception {
         doReturn(groups).when(connector).listConsumerGroups();
         doReturn(true).when(connector).shouldReplicateByTopicFilter(anyString());
         doReturn(true).when(connector).shouldReplicateByGroupFilter(anyString());
-        doReturn(offsets).when(connector).listConsumerGroupOffsets(anyString());
```
Comment (gharris1727) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnectorTest.java:154]: I think this leaves the `offsets` variable unused.

**Review by gharris1727 — COMMENTED** [review]
Hey @fonsdant Thanks for the PR, and thanks for contributing to Kafka!

Don't stress too much about the CI results, you should only be looking for build failures (the code doesn't compile) or consistent failures in the area you're changing. We have some flaky tests which cause a lot of noise in the results.

Overall this PR looks like it solves the problem I was concerned about, I just had a few suggestions.

**Comment by fonsdant** [comment]
@gharris1727, thank you for the appointed issues! As suggested, the public access modifier has been removed, a single groupSpec for map entries has been created, and tests have been mocked!

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnectorTest.java (hunk 4 above)
Comment (fonsdant) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnectorTest.java:154]: I have not used offsets anymore, but offsetsResult need to be mocked

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnector.java [hunk 5]
```diff
@@ -262,12 +264,13 @@ private void createInternalTopics() {
...
         return adminCall(
-                () -> sourceAdminClient.listConsumerGroupOffsets(group).partitionsToOffsetAndMetadata().get(),
-                () -> String.format("list offsets for consumer group %s on %s cluster", group,
-                        config.sourceClusterAlias())
+                () -> sourceAdminClient.listConsumerGroupOffsets(groupSpecs),
+                () -> String.format("list offsets for consumer groups %s on %s cluster", groups, config.sourceClusterAlias())
         );
```
Comment (gharris1727) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnector.java]: I think to make this adminCall functional, the first argument needs to be blocking. I think you need to use .all().get() instead of returning the whole Result object.

Comment (fonsdant) [reply connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnector.java]: While looping, we need to get topic partitions and metadata by group. For this reason, I have chosen `ListConsumerGroupOffsetsResult` as the return of listConsumerGroupOffsets as it allows us to do `result.partitionsToOffsetAndMetadata(group).get()` after calling `listConsumerGroups(filteredGroups)`.

In `.all().get()` inside `listConsumerGroups` approach, once we have all topic partitions and metadata for all consumer groups, we will not be able to distinguish each of them by group while looping.

An alternative could be to make a copy of `ListConsumerGroupOffsetsResult.all()` that groups future results by group. This way, we could continue to get them batched and still have how to distinguish what group each one belongs to. But I think it would need KIP, right? So it does not seem to be the case.

Let me know if I have missed something.

Many thanks for your review, @gharris1727! :)

Comment (gharris1727) [reply connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnector.java]: `ListConsumerGroupOffsetsResult#all()` has the return type `KafkaFuture<Map<String, Map<TopicPartition, OffsetAndMetadata>>>` which should include group information. I don't think it merges the results of the different groups together.

Comment (fonsdant) [reply connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnector.java]: Oh! You are right! Sorry, Greg, I have mixed up the return type. I will push a commit to fix it. Constructive review, thanks!

**Comment by fonsdant** [comment]
Hi, @gharris1727! Could you review it, please? :)

**Comment by mimaison** [comment]
@fonsdant Can you rebase the PR so it triggers the new CI? Thanks

**Comment by fonsdant** [comment]
@mimaison, done! Thanks :)

**Comment by mimaison** [comment]
Thanks for the update. Spotless is failing, see the error in https://github.com/apache/kafka/actions/runs/11364366312/job/31628998328?pr=17038

**Comment by fonsdant** [comment]
@mimaison, thanks! I have pushed an update for it.

**Review by mimaison — APPROVED** [review]
LGTM, thanks for the PR

**Comment by mimaison** [comment]
I'll leave the PR open for a couple of days in case @gharris1727 wants to take another look.

**Comment by fonsdant** [comment]
Thanks, @mimaison! 😄
